File: ios/App/App/Render/MotionMath.swift
import Foundation
import CoreGraphics

/// The Ken Burns and transition geometry for AVFoundation.
///
/// The Swift half of `src/lib/mobile/deviceRenderCaptions.ts` and the exact
/// counterpart of Android's `MotionMath.java`. The keyframe table is copied
/// verbatim from `remotion/montageMotion.ts` rather than paraphrased, because
/// three implementations of the same motion that were each "written from the
/// spec" would drift within a release.
///
/// COORDINATES. Remotion animates with a CSS transform:
///
///     transform: scale(s) translate(tx%, ty%);
///     transform-origin: focusX% focusY%;
///
/// CSS applies the list right-to-left, so a point p becomes
/// `o + s * (p + t - o)`, where t is in the element's own box. AVFoundation's
/// layer instructions take a `CGAffineTransform` in the composition's PIXEL
/// space with the origin at the bottom-left and y pointing UP, so the
/// conversion is: translate by the fraction times the canvas size (negating y),
/// scale, then correct for the origin.
enum MotionMath {

    struct Keyframes {
        let scaleFrom: CGFloat
        let scaleTo: CGFloat
        let translateXFrom: CGFloat
        let translateXTo: CGFloat
        let translateYFrom: CGFloat
        let translateYTo: CGFloat
    }

    /// Mirrors `KEN_BURNS_KEYFRAMES`. Translates are percentages.
    static func keyframes(_ motion: RenderManifest.Motion) -> Keyframes {
        switch motion {
        case .kenBurnsIn:
            return Keyframes(scaleFrom: 1.00, scaleTo: 1.25,
                             translateXFrom: 0, translateXTo: 0,
                             translateYFrom: 0, translateYTo: 0)
        case .kenBurnsOut:
            return Keyframes(scaleFrom: 1.25, scaleTo: 1.00,
                             translateXFrom: 0, translateXTo: 0,
                             translateYFrom: 0, translateYTo: 0)
        // The camera direction is opposite the image's translation: moving the
        // oversized image right reveals its left side, so the camera pans left.
        case .panLeft:
            return Keyframes(scaleFrom: 1.18, scaleTo: 1.18,
                             translateXFrom: -7, translateXTo: 7,
                             translateYFrom: 0, translateYTo: 0)
        case .panRight:
            return Keyframes(scaleFrom: 1.18, scaleTo: 1.18,
                             translateXFrom: 7, translateXTo: -7,
                             translateYFrom: 0, translateYTo: 0)
        case .staticFrame:
            return Keyframes(scaleFrom: 1.00, scaleTo: 1.00,
                             translateXFrom: 0, translateXTo: 0,
                             translateYFrom: 0, translateYTo: 0)
        }
    }

    static func clamp(_ value: CGFloat, _ minimum: CGFloat, _ maximum: CGFloat) -> CGFloat {
        Swift.min(maximum, Swift.max(minimum, value))
    }

    private static func lerp(_ from: CGFloat, _ to: CGFloat, _ progress: CGFloat) -> CGFloat {
        from + (to - from) * progress
    }

    /// Ken Burns state at `progress` (0...1 across the shot's slot).
    static func kenBurns(
        motion: RenderManifest.Motion, progress: CGFloat
    ) -> (scale: CGFloat, translateXFraction: CGFloat, translateYFraction: CGFloat) {
        let p = clamp(progress, 0, 1)
        let k = keyframes(motion)
        return (
            lerp(k.scaleFrom, k.scaleTo, p),
            lerp(k.translateXFrom, k.translateXTo, p) / 100,
            lerp(k.translateYFrom, k.translateYTo, p) / 100
        )
    }

    /// The scale-about-focus plus box-fraction translate, in canvas pixels.
    static func transform(
        scale: CGFloat,
        translateXFraction: CGFloat,
        translateYFraction: CGFloat,
        focusX: CGFloat,
        focusY: CGFloat,
        canvas: CGSize
    ) -> CGAffineTransform {
        // The focus point comes in as 0...1 from the top-left (CSS); AVFoundation
        // measures from the bottom-left.
        let originX = clamp(focusX, 0, 1) * canvas.width
        let originY = (1 - clamp(focusY, 0, 1)) * canvas.height

        // p + t, with y negated because CSS y grows downward.
        var matrix = CGAffineTransform(
            translationX: translateXFraction * canvas.width,
            y: -translateYFraction * canvas.height
        )
        // s * (p + t)
        matrix = matrix.concatenating(CGAffineTransform(scaleX: scale, y: scale))
        // + o * (1 - s)  ->  o + s * (p + t - o)
        matrix = matrix.concatenating(CGAffineTransform(
            translationX: originX * (1 - scale),
            y: originY * (1 - scale)
        ))
        return matrix
    }

    /// The extra flourish a named transition layers on the incoming shot while
    /// it dissolves in. `fade` is 0 at the start of the overlap and 1 when the
    /// shot is fully on screen.
    static func entrance(
        _ transition: RenderManifest.Transition, fade: CGFloat, canvas: CGSize
    ) -> CGAffineTransform {
        let f = clamp(fade, 0, 1)
        switch transition {
        case .slide:
            return transform(scale: 1, translateXFraction: (1 - f) * 0.12,
                             translateYFraction: 0, focusX: 0.5, focusY: 0.5, canvas: canvas)
        case .zoom:
            return transform(scale: 1.04 - 0.04 * f, translateXFraction: 0,
                             translateYFraction: 0, focusX: 0.5, focusY: 0.5, canvas: canvas)
        case .fade, .cut:
            return .identity
        }
    }

    /// `object-fit: cover` on the canvas: the scale that fills without
    /// letterboxing, and the offset that centres the overflow.
    static func coverFit(source: CGSize, canvas: CGSize) -> (scale: CGFloat, offset: CGPoint) {
        guard source.width > 0, source.height > 0 else { return (1, .zero) }
        let scale = Swift.max(canvas.width / source.width, canvas.height / source.height)
        return (
            scale,
            CGPoint(
                x: (canvas.width - source.width * scale) / 2,
                y: (canvas.height - source.height * scale) / 2
            )
        )
    }

    /// Where a picture goes on the canvas for a zoom and a focus point — the
    /// Swift copy of `framePlacement` in `src/lib/mobile/shotFraming.ts`.
    /// `frameZoom` 0 shows the whole picture (black fills the rest), 1 fills
    /// the canvas and crops around the focus point. The returned origin is
    /// top-left with y DOWN (a video composition's space); convert before
    /// drawing into a Core Graphics context.
    static func frameFit(
        source: CGSize,
        canvas: CGSize,
        frameZoom: CGFloat,
        focusX: CGFloat,
        focusY: CGFloat
    ) -> (scale: CGFloat, offset: CGPoint) {
        guard source.width > 0, source.height > 0 else { return (1, .zero) }
        let fit = Swift.min(canvas.width / source.width, canvas.height / source.height)
        let cover = Swift.max(canvas.width / source.width, canvas.height / source.height)
        let zoom = MotionMath.clamp(frameZoom, 0, 1)
        let scale = fit * pow(cover / fit, zoom)
        let width = source.width * scale
        let height = source.height * scale
        func place(_ drawn: CGFloat, _ room: CGFloat, _ focus: CGFloat) -> CGFloat {
            if drawn <= room { return (room - drawn) / 2 }
            return MotionMath.clamp(room / 2 - MotionMath.clamp(focus, 0, 1) * drawn, room - drawn, 0)
        }
        return (scale, CGPoint(x: place(width, canvas.width, focusX),
                               y: place(height, canvas.height, focusY)))
    }

    /// Playback rate for a clip whose slot outruns its footage — slow it down
    /// rather than freezing the last frame. Mirrors `computeClipPlaybackRate`.
    static func playbackRate(footageSeconds: Double, slotSeconds: Double) -> Double {
        guard footageSeconds > 0, slotSeconds > 0 else { return 1 }
        guard slotSeconds > footageSeconds else { return 1 }
        return Swift.min(1, Swift.max(0.8, footageSeconds / slotSeconds))
    }
}

File: android/app/src/main/java/com/rclipper/app/render/MotionMath.java
package com.rclipper.app.render;

import android.graphics.Matrix;

/**
 * The Ken Burns and transition geometry, in the form Media3 wants it.
 *
 * This is the Java half of `src/lib/mobile/deviceRenderCaptions.ts`, which is
 * itself the phone-side restatement of `remotion/montageMotion.ts`. All three
 * have to agree, so the keyframe table below is copied verbatim rather than
 * paraphrased, and the only interesting code here is the coordinate conversion.
 *
 * COORDINATES. Remotion animates with a CSS transform:
 *
 *     transform: scale(s) translate(tx%, ty%);
 *     transform-origin: focusX% focusY%;
 *
 * CSS applies a transform list right-to-left, so a point p becomes
 * `o + s * (p + t - o)` where `t` is the translate expressed in the element's
 * own box and `o` is the origin. Media3's {@code MatrixTransformation} works in
 * normalised device coordinates: the frame spans [-1, 1] on both axes, y points
 * UP (CSS y points down), and the matrix maps input vertices to output
 * vertices. Converting is therefore:
 *
 *     t_ndc = (fraction * 2, -fraction * 2)      // box width is 2 in NDC
 *     o_ndc = (focusX * 2 - 1, 1 - focusY * 2)
 *     M = translate(t_ndc) then scale(s) then translate(o_ndc * (1 - s))
 *
 * Getting this wrong is the classic phone-render bug: the zoom looks right and
 * the pans drift the wrong way or by twice the distance, which nobody notices
 * until a side-by-side comparison.
 */
public final class MotionMath {

    private MotionMath() {}

    /** Mirrors `KEN_BURNS_KEYFRAMES`. */
    public static float[] keyframes(String motion) {
        // { scaleFrom, scaleTo, txFrom, txTo, tyFrom, tyTo } — translates in percent.
        switch (motion == null ? "static" : motion) {
            case "ken_burns_in":  return new float[] { 1.00f, 1.25f,  0f,  0f, 0f, 0f };
            case "ken_burns_out": return new float[] { 1.25f, 1.00f,  0f,  0f, 0f, 0f };
            // The camera direction is opposite the image's translation: moving
            // the oversized image right reveals its left side, so the camera
            // pans left.
            case "pan_left":      return new float[] { 1.18f, 1.18f, -7f,  7f, 0f, 0f };
            case "pan_right":     return new float[] { 1.18f, 1.18f,  7f, -7f, 0f, 0f };
            default:              return new float[] { 1.00f, 1.00f,  0f,  0f, 0f, 0f };
        }
    }

    public static float clamp(float value, float min, float max) {
        return Math.min(max, Math.max(min, value));
    }

    private static float lerp(float from, float to, float progress) {
        return from + (to - from) * progress;
    }

    /**
     * The Ken Burns matrix for a still at `progress` (0..1 across its slot),
     * about the subject-focus origin.
     */
    public static Matrix kenBurns(String motion, float progress, float focusX, float focusY) {
        float p = clamp(progress, 0f, 1f);
        float[] k = keyframes(motion);
        float scale = lerp(k[0], k[1], p);
        float txFraction = lerp(k[2], k[3], p) / 100f;
        float tyFraction = lerp(k[4], k[5], p) / 100f;
        return transform(scale, txFraction, tyFraction, focusX, focusY);
    }

    /**
     * Build the NDC matrix for a scale about a focus origin plus a translate
     * expressed as a fraction of the element's own box.
     */
    public static Matrix transform(
        float scale, float txFraction, float tyFraction, float focusX, float focusY
    ) {
        float originX = clamp(focusX, 0f, 1f) * 2f - 1f;
        // CSS y grows downward, NDC y grows upward.
        float originY = 1f - clamp(focusY, 0f, 1f) * 2f;

        Matrix matrix = new Matrix();
        // p + t
        matrix.postTranslate(txFraction * 2f, -tyFraction * 2f);
        // s * (p + t)
        matrix.postScale(scale, scale);
        // + o * (1 - s)  →  o + s * (p + t - o)
        matrix.postTranslate(originX * (1f - scale), originY * (1f - scale));
        return matrix;
    }

    /**
     * The extra flourish a named transition layers on top of the dissolve, as a
     * matrix for the INCOMING shot during its fade window.
     *
     *   slide — `translateX((1 - fade) * 12%)`
     *   zoom  — `scale(1.04 - 0.04 * fade)`
     *   fade / cut — none
     *
     * `fade` is the dissolve progress, 0 at the start of the overlap and 1 when
     * the incoming shot is fully on screen.
     */
    public static Matrix entrance(String transition, float fade) {
        float f = clamp(fade, 0f, 1f);
        if ("slide".equals(transition)) {
            return transform(1f, (1f - f) * 0.12f, 0f, 0.5f, 0.5f);
        }
        if ("zoom".equals(transition)) {
            return transform(1.04f - 0.04f * f, 0f, 0f, 0.5f, 0.5f);
        }
        return new Matrix();
    }

    /**
     * Playback rate for a clip whose slot outruns its footage — slow it down
     * rather than freezing the last frame. Mirrors `computeClipPlaybackRate`;
     * `MIN_CLIP_PLAYBACK_RATE` is 0.8, i.e. at most a 1.25x stretch.
     */
    public static float playbackRate(double footageSeconds, double slotSeconds) {
        if (!(footageSeconds > 0) || !(slotSeconds > 0)) return 1f;
        if (slotSeconds <= footageSeconds) return 1f;
        return clamp((float) (footageSeconds / slotSeconds), 0.8f, 1f);
    }

    /**
     * Allocate whole frames across a scene's shots in proportion to their
     * durations, every shot getting at least one frame and the last absorbing
     * the remainder. Mirrors `allocateAssetFrames` so a scene is the same
     * length on all three renderers.
     *
     * @return {from, durationInFrames} pairs, contiguous and covering [0, total).
     */
    public static int[][] allocateFrames(double[] durationsSeconds, int totalFrames) {
        int n = durationsSeconds.length;
        if (n == 0) return new int[0][];

        int safeTotal = Math.max(n, totalFrames);
        double sum = 0;
        for (double d : durationsSeconds) if (d > 0) sum += d;

        int[][] ranges = new int[n][2];
        int acc = 0;
        for (int i = 0; i < n; i++) {
            double weight = sum > 0 ? Math.max(0, durationsSeconds[i]) / sum : 1d / n;
            int frames;
            if (i == n - 1) {
                frames = Math.max(1, safeTotal - acc);
            } else {
                int remainingAfter = n - 1 - i;
                int maxForThis = Math.max(1, safeTotal - acc - remainingAfter);
                frames = Math.min(maxForThis, Math.max(1, (int) Math.round(weight * safeTotal)));
            }
            ranges[i][0] = acc;
            ranges[i][1] = frames;
            acc += frames;
        }
        return ranges;
    }
}
